ravenframework/CustomModes/SlurmSimulationMode.py

```diff
@@ -0,0 +1,216 @@
+      #the batchsize is just the number of nodes of which there is one
+      # per line in the nodeFile divided by the numMPI (which is per run)
+      # and the floor and int and max make sure that the numbers are reasonable
+      maxBatchsize = max(int(math.floor(len(lines) / numMPI)), 1)
+
+      if maxBatchsize < oldBatchsize:
+        newRunInfo['batchSize'] = maxBatchsize
+        self.raiseAWarning("changing batchsize from "+str(oldBatchsize)+" to "+str(maxBatchsize)+" to fit on "+str(len(lines))+" processors")
+      newBatchsize = newRunInfo['batchSize']
+      self.raiseADebug('Batch size is "{}"'.format(newBatchsize))
+      if newBatchsize > 1:
+        #need to split node lines so that numMPI nodes are available per run
+        workingDir = runInfoDict['WorkingDir']
+        for i in range(newBatchsize):
+          nodeFile = open(os.path.join(workingDir, f"node_{i}"), "w")
+          for line in lines[i*numMPI : (i+1) * numMPI]:
+            nodeFile.write(line)
+          nodeFile.close()
+        #then give each index a separate file.
+        nodeCommand = runInfoDict["NodeParameter"]+" %BASE_WORKING_DIR%/node_%INDEX% "
+      else:
+        #If only one batch just use original node file
+        nodeCommand = runInfoDict["NodeParameter"]+" "+nodeFile
+
+    else:
+      #Not in PBS, so can't look at PBS_NODEFILE and none supplied in input
+      newBatchsize = newRunInfo['batchSize']
+      numMPI = runInfoDict['NumMPI']
+      #TODO, we don't have a way to know which machines it can run on
+      # when not in PBS so just distribute it over the local machine:
+      nodeCommand = " "
+
+    if len(self.__mpiparams) > 0:
+      mpiParams = " ".join(self.__mpiparams)+" "
+    else:
+      mpiParams = ""
+    # Create the mpiexec pre command
+    # Note, with defaults the precommand is "mpiexec -f nodeFile -n numMPI"
+    if self.__createPrecommand:
+      newRunInfo['precommand'] = runInfoDict["MPIExec"]+" "+mpiParams+nodeCommand+" -n "+str(numMPI)+" "+runInfoDict['precommand']
+    else:
+      newRunInfo['precommand'] = runInfoDict['precommand']
+    if runInfoDict['NumThreads'] > 1:
+      newRunInfo['threadParameter'] = runInfoDict['threadParameter']
+      #add number of threads to the post command.
+      newRunInfo['postcommand'] =" {} {}".format(newRunInfo['threadParameter'],runInfoDict['postcommand'])
+    self.raiseAMessage("precommand: "+newRunInfo['precommand']+", postcommand: "+newRunInfo.get('postcommand',runInfoDict['postcommand']))
+    return newRunInfo
+
+  def __createAndRunSbatch(self, runInfoDict):
+    """
+      Generates a SLURM sbatch command to run the simulation
+      @ In, runInfoDict, dict, dictionary of run info.
+      @ Out, remoteRunCommand, dict, dictionary of command.
+    """
+    # determine the cores needed for the job
+    if self.__coresNeeded is not None:
+      coresNeeded = self.__coresNeeded
+    else:
+      coresNeeded = runInfoDict['batchSize']*runInfoDict['NumMPI']
```

Comment:
NumThreads needs to be treated specially because those need to have shared memory. So we calculate coresNeeded and ncpus and make the sbatch (or qsub in for PBS) command request so that the ncpus are on the same node and share memory.  I will add some comments about this in the code.